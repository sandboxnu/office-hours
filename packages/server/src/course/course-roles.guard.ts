import { Injectable, UnauthorizedException } from '@nestjs/common';
import { UserModel } from '../profile/user.entity';
import { RolesGuard } from '../guards/role.guard';

@Injectable()
export class CourseRolesGuard extends RolesGuard {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  async setupData(
    request: any,
  ): Promise<{ courseId: number; user: UserModel }> {
    const user = await UserModel.findOne(request.user.userId, {
      relations: ['courses'],
    });

    const courseId = request.params.id;
    return { courseId, user };
  }
}
