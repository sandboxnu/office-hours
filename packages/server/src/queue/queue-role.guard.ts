import { Injectable, UnauthorizedException } from '@nestjs/common';
import { UserModel } from '../profile/user.entity';
import { RolesGuard } from '../guards/role.guard';
import { QueueModel } from './queue.entity';

@Injectable()
export class QueueRolesGuard extends RolesGuard {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  async setupData(
    request: any,
  ): Promise<{ courseId: number; user: UserModel }> {
    const queue = await QueueModel.findOne(request.params.queueId);
    const courseId = queue?.courseId;
    const user = await UserModel.findOne(request.user.userId, {
      relations: ['courses'],
    });

    return { courseId, user };
  }
}
