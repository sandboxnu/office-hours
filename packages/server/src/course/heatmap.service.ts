import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';

import { Cache } from 'cache-manager';
import { ClosedQuestionStatus, Heatmap, timeDiffInMins } from '@koh/common';
import moment = require('moment');
import { CourseModel } from './course.entity';
import { MoreThan } from 'typeorm';
import { QuestionModel } from '../question/question.entity';
import { EventModel } from '../profile/event-model.entity';
import { Command, Positional } from 'nestjs-command';
import { inRange, mean, range } from 'lodash';
import 'moment-timezone';

function arrayRotate(arr, count) {
  count -= arr.length * Math.floor(count / arr.length);
  const spliced = arr.splice(0, count);
  return [...arr, ...spliced];
}

@Injectable()
export class HeatmapService {
  constructor(@Inject(CACHE_MANAGER) private cacheManager: Cache) {}

  async getCachedHeatmapFor(courseId: number): Promise<Heatmap | false> {
    //One week
    const cacheLengthInSeconds = 604800;
    return this.cacheManager.wrap(
      `heatmap/${courseId}`,
      () => this._getHeatmapFor(courseId),
      { ttl: cacheLengthInSeconds },
    );
  }

  // Do not use this externally plz
  async _getHeatmapFor(courseId: number): Promise<Heatmap | false> {
    // The number of minutes to average across
    const BUCKET_SIZE_IN_MINS = 15;
    // Number of samples to gather per bucket
    const SAMPLES_PER_BUCKET = 3;
    console.time('heatmap');
    const recent = moment().subtract(8, 'weeks').toISOString();
    const questions = await QuestionModel.createQueryBuilder('question')
      .leftJoinAndSelect('question.queue', 'queue')
      .where('queue.courseId = :courseId', { courseId })
      .andWhere('question.status = :status', {
        status: ClosedQuestionStatus.Resolved,
      })
      .andWhere('question.helpedAt IS NOT NULL')
      .andWhere('question.createdAt > :recent', { recent })
      .orderBy('question.createdAt', 'ASC')
      .getMany();
    if (questions.length === 0) {
      return false;
    }

    const taEvents = await EventModel.find({
      where: { time: MoreThan(recent), courseId },
    });

    if (taEvents.length === 0) {
      return false;
    }

    const tz = (await CourseModel.findOne({ id: courseId })).timezone;

    function extractTimestamps(taEvents: EventModel[]) {
      const hours = [];
      taEvents.sort((a, b) => {
        return a.time.getTime() - b.time.getTime();
      });

      let iday = taEvents[0].time.getDate();
      let itime = taEvents[0].time.getTime();
      let etime = taEvents[0].time.getTime();
      for (let i = 1; i < taEvents.length; i++) {
        if (taEvents[i].time.getDate() == iday) {
          etime = taEvents[i].time.getTime();
        } else {
          hours.push([itime, etime]);
          iday = taEvents[i].time.getDate();
          itime = taEvents[i].time.getTime();
          etime = taEvents[i].time.getTime(); // new day
        }
      }

      hours.push([itime, etime]); // push last day.

      return hours;
    }

    let heatmap = this._generateHeatMapWithReplay(
      // Ignore questions that cross midnight (usually a fluke)
      questions.filter((q) => q.helpedAt.getDate() === q.createdAt.getDate()),
      extractTimestamps(taEvents),
      tz,
      BUCKET_SIZE_IN_MINS,
      SAMPLES_PER_BUCKET,
    );
    heatmap = arrayRotate(
      heatmap,
      -moment.tz.zone(tz).utcOffset(Date.now()) / BUCKET_SIZE_IN_MINS,
    );
    console.timeEnd('heatmap');
    return heatmap;
  }

  // PRIVATE function that is public for testing purposes
  // Rewind through the last few weeks and for each time interval,
  // figure out how long wait time would have been if you had joined the queue at that time
  // Timezone should be IANA
  // Returns heatmap in the timezone (ie 3rd bucket is 3am in that timezone)
  _generateHeatMapWithReplay(
    questions: QuestionModel[],
    hourTimestamps: [number, number][],
    timezone: string,
    bucketSize: number,
    samplesPerBucket: number,
  ): Heatmap {
    const sampleInterval = bucketSize / samplesPerBucket;
    /*
    TEST: Question1 is  3:05 - 3:25
    // The next question is 3:21 - 3:49
    THe following question is 4:05 - 4:10
    
    Bucket = 60, Samples = 3, so timepoints are: 3:00, 3:20, 3:40.

    3:20 sample gets waittime of 5 minutes
    3:40 samples get waittimes of 9 minutes
    4:00 sample gets waittime of 0 minutes


    If i entered the queue at that time when should I have gotten help?
    Every interval of minutes for the past 5 weeks are aggregated (by taking the avg)
    
    analyze the buckets to find the closest time approximation

    look at question Q1 and the next question Q2
    for all sample timepoints between Q1.createdAt and Q2.createdAt:
       - sample = Q1.helpedAt - timepoint (if negative, then it's 0)

  */

    function dateToBucket(date: Date | number): number {
      // parse in zone to handle daylight savings by getting day/hour/minute within that IANA zone
      const cInZone = moment.tz(date, timezone);
      return Math.floor(
        (cInZone.day() * 24 * 60 + cInZone.hour() * 60 + cInZone.minute()) /
          bucketSize,
      );
    }
    const timepointBuckets: number[][] = [
      ...Array((24 * 7 * 60) / bucketSize),
    ].map(() => []);

    if (questions.length) {
      const startDate = questions[0].createdAt;
      const sunday = moment.tz(startDate, timezone).startOf('week').toDate();

      function getNextTimepointIndex(date: Date): number {
        return Math.floor(timeDiffInMins(date, sunday) / sampleInterval) + 1;
      }

      // Get the date of the sample timepoint immediately after the given date
      function getNextSampleTimepoint(date: Date): Date {
        const timepointIndex = getNextTimepointIndex(date);
        return new Date(
          sunday.getTime() + timepointIndex * sampleInterval * 60 * 1000,
        );
      }

      // Get all timepoints between the two dates
      function getSampleTimepointsInDateRange(
        date1: Date,
        date2: Date,
      ): Date[] {
        const ret = [];
        let curr = getNextSampleTimepoint(date1);
        while (curr.getTime() < date2.getTime()) {
          ret.push(curr);
          curr = getNextSampleTimepoint(curr);
        }
        return ret;
      }

      // Get the start time of the current bucket
      function lastBucketBoundary(date: Date): moment.Moment {
        const startOfWeek = moment.tz(date, timezone).startOf('week');
        const m = moment(date);
        return m.subtract(m.diff(startOfWeek, 'm') % bucketSize, 'm');
      }

      // go two questions at a time
      let isFirst = true;
      for (let i = 0; i < questions.length; i++) {
        const curr = questions[i];
        const next = questions[i + 1];
        const isLast = i === questions.length - 1;

        // get the timepoints in between
        let sampledTimepoints = getSampleTimepointsInDateRange(
          isFirst
            ? lastBucketBoundary(curr.createdAt)
                .subtract(1, 's') // so that we get the first timepoint
                .toDate()
            : curr.createdAt,
          isLast
            ? lastBucketBoundary(curr.helpedAt)
                .add(bucketSize, 'm') // to get the nextBucketBoundary
                .toDate()
            : next.createdAt,
        );
        sampledTimepoints = sampledTimepoints.filter((time) =>
          hourTimestamps.some(([start, end]) =>
            inRange(time.getTime(), start, end),
          ),
        );

        // Pad the first bucket with zeros to account for timepoints before the first
        if (sampledTimepoints.length > 0 && isFirst) {
          isFirst = false;
        }
        // When we would have hypothetically gotten help at this timepoint
        for (const c of sampledTimepoints) {
          let wait = 0;
          if (
            inRange(
              c.getTime(),
              curr.createdAt.getTime(),
              curr.helpedAt.getTime(),
            )
          ) {
            wait = (curr.helpedAt.getTime() - c.getTime()) / 60000;
          }

          const bucketIndex = dateToBucket(c);
          timepointBuckets[bucketIndex].push(wait);
        }
      }
    }

    // Were there ever office hours in this bucket?
    const wereHoursDuringBucket: boolean[] = [
      ...Array((24 * 7 * 60) / bucketSize),
    ];
    for (const [start, end] of hourTimestamps) {
      //prevents an office hour from [N, M] to register in multiple buckets
      for (const i of range(dateToBucket(start), dateToBucket(end - 1) + 1)) {
        wereHoursDuringBucket[i] = true;
      }
    }

    const h: Heatmap = timepointBuckets.map((samples, i) => {
      if (samples.length > 0) {
        return mean(samples);
      } else if (wereHoursDuringBucket[i]) {
        return 0;
      } else {
        return -1;
      }
    });
    return h;
  }

  @Command({
    command: 'heatmap:generate <courseId>',
    describe: 'generate heatmap for a course',
    autoExit: true,
  })
  async create(
    @Positional({
      name: 'courseId',
      describe: 'which course the heatmap will be generated for',
      type: 'number',
    })
    courseId: number,
  ): Promise<void> {
    console.log(await this._getHeatmapFor(courseId));
  }
}
