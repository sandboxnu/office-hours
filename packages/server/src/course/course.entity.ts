import { Heatmap } from '@koh/common';
import { AsyncQuestionModel } from '../asyncQuestion/asyncQuestion.entity';
import { Exclude } from 'class-transformer';
import { QuestionTypeModel } from 'question/question-type.entity';
import {
  BaseEntity,
  Column,
  DeleteDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { AlertModel } from '../alerts/alerts.entity';
import { EventModel } from '../profile/event-model.entity';
import { UserCourseModel } from '../profile/user-course.entity';
import { QueueModel } from '../queue/queue.entity';
import { SemesterModel } from '../semester/semester.entity';
import { OrganizationCourseModel } from '../organization/organization-course.entity';
@Entity('course_model')
export class CourseModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @OneToMany((type) => QueueModel, (q) => q.course)
  queues: QueueModel[];

  @Column('text')
  name: string; // display name entered by Prof

  @Column('text')
  sectionGroupName: string; // from admin

  @Column('text', { nullable: true })
  coordinator_email: string;

  @Column('text', { nullable: true })
  @Exclude()
  icalURL: string;

  @Column('text', { nullable: true })
  @Exclude()
  zoomLink: string;

  @Column('integer', { nullable: true })
  @Exclude()
  questionTimer: number;

  @OneToMany((type) => UserCourseModel, (ucm) => ucm.course)
  @Exclude()
  userCourses: UserCourseModel;

  @ManyToOne((type) => SemesterModel, (semester) => semester.courses)
  @JoinColumn({ name: 'semesterId' })
  @Exclude()
  semester: SemesterModel;

  @Column({ nullable: true })
  @Exclude()
  // TODO: can we make these not nullable and work with TypeORM
  semesterId: number;

  @Column('boolean', { nullable: true })
  enabled: boolean; // Set to true if the given the course is using our app

  // The heatmap is false when there havent been any questions asked yet or there havent been any office hours
  heatmap: Heatmap | false;

  // The IANA string representing the timezone the course is centered around. This is important for any time based events for a course
  @Column('text', { nullable: true })
  timezone: string;

  @OneToMany((type) => EventModel, (event) => event.course)
  @Exclude()
  events: EventModel[];

  @OneToMany((type) => AlertModel, (alert) => alert.course)
  @Exclude()
  alerts: AlertModel[];

  // Whether or not students are allowed to self-enroll in the class
  // WARNING: THIS SHOULD ONLY BE USED AS A TEMPORARY MEASURE WHEN THINGS LIKE BANNER ARE DOWN
  @Column('boolean', { nullable: true, default: false })
  selfEnroll: boolean;

  @OneToMany(
    (type) => AsyncQuestionModel,
    (asyncQuestion) => asyncQuestion.course,
  )
  @Exclude()
  images: AsyncQuestionModel[];

  @Column('text', { array: true, nullable: true, default: [] })
  asyncQuestionDisplayTypes: string[];

  @DeleteDateColumn()
  deletedAt?: Date;

  @ManyToOne(
    (type) => OrganizationCourseModel,
    (organizationCourse) => organizationCourse.course,
  )
  @Exclude()
  organizationCourse: OrganizationCourseModel;

  @Column({ nullable: true })
  organizationCourseId: number;
}
