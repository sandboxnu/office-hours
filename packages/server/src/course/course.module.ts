import { CacheModule, Module } from '@nestjs/common';
import { QueueModule } from '../queue/queue.module';
import { LoginModule } from '../login/login.module';
import { LoginCourseService } from '../login/login-course.service';
import { CourseController } from './course.controller';
import { CourseService } from './course.service';
import { HeatmapService } from './heatmap.service';

@Module({
  controllers: [CourseController],
  imports: [QueueModule, LoginModule, CacheModule.register()],
  providers: [LoginCourseService, HeatmapService, CourseService],
})
export class CourseModule {}
