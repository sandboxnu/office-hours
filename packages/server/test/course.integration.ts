import {
  ERROR_MESSAGES,
  Role,
  TACheckinTimesResponse,
  TACheckoutResponse,
} from '@koh/common';
import { EventModel, EventType } from 'profile/event-model.entity';
import { UserCourseModel } from 'profile/user-course.entity';
import { CourseModule } from '../src/course/course.module';
import { QueueModel } from '../src/queue/queue.entity';
import {
  ClosedOfficeHourFactory,
  CourseFactory,
  EventFactory,
  OfficeHourFactory,
  QuestionFactory,
  QueueFactory,
  StudentCourseFactory,
  TACourseFactory,
  UserCourseFactory,
  UserFactory,
} from './util/factories';
import { setupIntegrationTest } from './util/testUtils';

describe('Course Integration', () => {
  const supertest = setupIntegrationTest(CourseModule);
  describe('GET /courses/:id', () => {
    it('gets office hours no queues, since no queue is happening right now', async () => {
      const course = await CourseFactory.create({
        officeHours: [await ClosedOfficeHourFactory.create()],
        timezone: 'America/New_York',
      });
      await QueueFactory.create();

      await UserCourseFactory.create({
        user: await UserFactory.create(),
        course: course,
      });
      // will not load b/c office hours aren't happening right now
      // (unless you go back in time and run these tests )
      const response = await supertest({ userId: 1 })
        .get(`/courses/${course.id}`)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('gets office hours and queues since time is now and rooms are same', async () => {
      const now = new Date();
      const course = await CourseFactory.create();

      await QueueFactory.create({
        room: "Matthias's Office",
        course: course,
        officeHours: [
          await OfficeHourFactory.create({
            startTime: now,
            endTime: new Date(now.valueOf() + 4500000),
            room: "Matthias's Office",
          }),
          await OfficeHourFactory.create(), // aren't loaded cause time off
        ],
      });

      await QueueFactory.create({
        course: course,
      });

      await UserCourseFactory.create({
        user: await UserFactory.create(),
        course: course,
      });

      const response = await supertest({ userId: 1 })
        .get(`/courses/${course.id}`)
        .expect(200);

      expect(response.body).toMatchObject({
        queues: [{ id: 1 }, { id: 2 }],
      });
    });

    it('cant get office hours if not a member of the course', async () => {
      const now = new Date();
      const course = await CourseFactory.create();

      await QueueFactory.create({
        room: "Matthias's Office",
        course: course,
        officeHours: [
          await OfficeHourFactory.create({
            startTime: now,
            endTime: new Date(now.valueOf() + 4500000),
            room: "Matthias's Office",
          }),
          await OfficeHourFactory.create(), // aren't loaded cause time off
        ],
      });

      await supertest({ userId: 1 }).get(`/courses/${course.id}`).expect(401);
    });
  });

  //TODO: make a DSL for testing auth points using Hack your own Language
  describe('POST /courses/:id/ta_location/:room', () => {
    it('checks a TA into an existing queue', async () => {
      const queue = await QueueFactory.create();
      const ta = await UserFactory.create();
      await TACourseFactory.create({
        course: queue.course,
        user: ta,
      });

      const response = await supertest({ userId: ta.id })
        .post(`/courses/${queue.course.id}/ta_location/${queue.room}`)
        .expect(201);

      expect(response.body).toMatchSnapshot();

      const events = await EventModel.find();
      expect(events.length).toBe(1);
      expect(events[0].eventType).toBe(EventType.TA_CHECKED_IN);
      expect(events[0].queueId).toBe(queue.id);
    });

    it("Doesn't allow student to check in", async () => {
      const queue = await QueueFactory.create();
      const student = await UserFactory.create();
      await StudentCourseFactory.create({
        course: queue.course,
        user: student,
      });

      await supertest({ userId: student.id })
        .post(`/courses/${queue.course.id}/ta_location/${queue.room}`)
        .expect(401);

      const events = await EventModel.find();
      expect(events.length).toBe(0);
    });

    it("Doesn't allow a TA to check into a new queue", async () => {
      const ta = await UserFactory.create();
      const tcf = await TACourseFactory.create({
        course: await CourseFactory.create(),
        user: ta,
      });
      await supertest({ userId: ta.id })
        .post(`/courses/${tcf.courseId}/ta_location/The Alamo`)
        .expect(403);

      const events = await EventModel.find();
      expect(events.length).toBe(0);
    });

    it('Allows a professor to create a new queue', async () => {
      const professor = await UserFactory.create();
      const pcf = await UserCourseFactory.create({
        course: await CourseFactory.create(),
        user: professor,
        role: Role.PROFESSOR,
      });
      const response = await supertest({ userId: professor.id })
        .post(`/courses/${pcf.courseId}/ta_location/The Alamo`)
        .expect(201);

      expect(response.body).toMatchObject({
        id: 1,
        room: 'The Alamo',
        staffList: [{ id: professor.id }],
      });

      const events = await EventModel.find();
      expect(events.length).toBe(1);
      expect(events[0].eventType).toBe(EventType.TA_CHECKED_IN);
    });

    it("Doesn't allow users to check into multiple queues", async () => {
      const queue1 = await QueueFactory.create();
      const ta = await UserFactory.create();
      await TACourseFactory.create({
        course: queue1.course,
        user: ta,
      });
      const queue2 = await QueueFactory.create({
        course: queue1.course,
      });

      await supertest({ userId: ta.id })
        .post(`/courses/${queue1.courseId}/ta_location/${queue1.room}`)
        .expect(201);

      let events = await EventModel.count();
      expect(events).toBe(1);

      const response2 = await supertest({ userId: ta.id })
        .post(`/courses/${queue2.courseId}/ta_location/${queue2.room}`)
        .expect(401);
      expect(response2.body.message).toBe(
        ERROR_MESSAGES.courseController.checkIn.cannotCheckIntoMultipleQueues,
      );
      events = await EventModel.count();
      expect(events).toBe(1);
    });
  });

  describe('DELETE, /courses/:id/ta_location/:room', () => {
    it('tests TA is checked out from queue if exists', async () => {
      const ta = await UserFactory.create();
      const queue = await QueueFactory.create({
        room: 'The Alamo',
        staffList: [ta],
      });
      const tcf = await TACourseFactory.create({
        course: queue.course,
        user: ta,
      });

      expect(
        (await QueueModel.findOne({}, { relations: ['staffList'] })).staffList
          .length,
      ).toEqual(1);

      await supertest({ userId: ta.id })
        .delete(`/courses/${tcf.courseId}/ta_location/The Alamo`)
        .expect(200);

      expect(
        await QueueModel.findOne({}, { relations: ['staffList'] }),
      ).toMatchObject({
        staffList: [],
      });

      const events = await EventModel.find();
      expect(events.length).toBe(1);
      expect(events[0].eventType).toBe(EventType.TA_CHECKED_OUT);
      expect(events[0].queueId).toBe(queue.id);
    });

    it('tests student cant checkout from queue', async () => {
      const student = await UserFactory.create();
      const queue = await QueueFactory.create({
        room: 'The Alamo',
      });
      const scf = await StudentCourseFactory.create({
        course: queue.course,
        user: student,
      });

      await supertest({ userId: student.id })
        .delete(`/courses/${scf.courseId}/ta_location/The Alamo`)
        .expect(401);
    });

    it('returns canClearQueue true when TA checks out', async () => {
      const ofs = await ClosedOfficeHourFactory.create();
      const ta = await UserFactory.create();
      const queue = await QueueFactory.create({
        room: 'The Alamo',
        staffList: [ta],
        officeHours: [ofs],
      });
      const tcf = await TACourseFactory.create({
        course: queue.course,
        user: ta,
      });
      await QuestionFactory.create({ queue: queue });

      const checkoutResult: TACheckoutResponse = (
        await supertest({ userId: ta.id })
          .delete(`/courses/${tcf.courseId}/ta_location/The Alamo`)
          .expect(200)
      ).body;

      expect(checkoutResult.canClearQueue).toBeTruthy();
    });

    it('tests nothing happens if ta not in queue', async () => {
      const ta = await UserFactory.create();
      const queue = await QueueFactory.create({ room: 'The Alamo' });
      const tcf = await TACourseFactory.create({
        course: queue.course,
        user: ta,
      });

      await supertest({ userId: ta.id })
        .delete(`/courses/${tcf.courseId}/ta_location/The Alamo`)
        .expect(200);

      expect(
        await QueueModel.findOne({}, { relations: ['staffList'] }),
      ).toMatchObject({
        staffList: [],
      });

      const UCM = UserCourseModel.findOne({
        where: { userId: ta.id, courseId: queue.courseId },
      });
      const events = await EventModel.find({
        where: { user: UCM },
      });

      expect(events.length).toBe(0);
    });
  });

  describe('POST /courses/:id/update_override', () => {
    it('tests creating override using the endpoint', async () => {
      const course = await CourseFactory.create();
      const user = await UserFactory.create();
      const professor = await UserFactory.create();
      await UserCourseFactory.create({
        user: professor,
        role: Role.PROFESSOR,
        course,
      });
      await supertest({ userId: professor.id })
        .post(`/courses/${course.id}/update_override`)
        .send({ email: user.email, role: Role.STUDENT })
        .expect(201);
      const ucm = await UserCourseModel.findOne({
        where: {
          userId: user.id,
        },
      });
      expect(ucm.role).toEqual(Role.STUDENT);
      expect(ucm.override).toBeTruthy();
    });
  });

  describe('DELETE /courses/:id/update_override', () => {
    it('tests deleting override using the endpoint', async () => {
      const course = await CourseFactory.create();
      const user = await UserFactory.create();
      const professor = await UserFactory.create();
      await UserCourseFactory.create({
        user: professor,
        role: Role.PROFESSOR,
        course,
      });
      await UserCourseFactory.create({
        user: user,
        role: Role.TA,
        override: true,
        course,
      });

      await supertest({ userId: professor.id })
        .delete(`/courses/${course.id}/update_override`)
        .send({ email: user.email, role: Role.STUDENT })
        .expect(200);

      const ucm = await UserCourseModel.findOne({
        where: {
          userId: user.id,
        },
      });
      expect(ucm).toBeUndefined();
    });
  });

  describe('GET /courses/:id/ta_check_in_times', () => {
    it('tests that events within date range are gotten', async () => {
      const now = new Date();
      const yesterday = new Date();
      yesterday.setUTCHours(now.getUTCHours() - 24);

      const course = await CourseFactory.create();
      const ta = await UserFactory.create();
      const professor = await UserFactory.create();

      await UserCourseFactory.create({
        user: ta,
        role: Role.TA,
        course,
      });
      await UserCourseFactory.create({
        user: professor,
        role: Role.PROFESSOR,
        course,
      });

      await EventFactory.create({
        user: ta,
        course: course,
        time: yesterday,
        eventType: EventType.TA_CHECKED_IN,
      });

      const yesterdayPlusTwoHours = new Date(yesterday);
      yesterdayPlusTwoHours.setUTCHours(yesterday.getUTCHours() + 2);

      await EventFactory.create({
        user: ta,
        course: course,
        time: new Date(yesterdayPlusTwoHours),
        eventType: EventType.TA_CHECKED_OUT,
      });

      const thenThreeMoreHours = new Date(yesterdayPlusTwoHours);
      thenThreeMoreHours.setUTCHours(yesterday.getUTCHours() + 3);

      await EventFactory.create({
        user: ta,
        course: course,
        time: thenThreeMoreHours,
        eventType: EventType.TA_CHECKED_IN,
      });

      const twelveHoursAFter = new Date(thenThreeMoreHours);
      twelveHoursAFter.setUTCHours(thenThreeMoreHours.getUTCHours() + 12);

      await EventFactory.create({
        user: ta,
        course: course,
        time: twelveHoursAFter,
        eventType: EventType.TA_CHECKED_OUT_FORCED,
      });

      const justNow = new Date(Date.now() - 1000);

      await EventFactory.create({
        user: ta,
        course: course,
        time: justNow,
        eventType: EventType.TA_CHECKED_IN,
      });

      const twoDaysAgo = new Date();
      twoDaysAgo.setUTCDate(twoDaysAgo.getUTCDate() - 2);

      const data = await supertest({ userId: professor.id })
        .get(`/courses/${course.id}/ta_check_in_times`)
        .query({
          startDate: twoDaysAgo,
          endDate: new Date(),
        })
        .expect(200);

      const checkinTimes = (data.body as unknown as TACheckinTimesResponse)
        .taCheckinTimes;

      const taName = ta.firstName + ' ' + ta.lastName;

      expect(checkinTimes.length).toBe(3);
      expect(checkinTimes).toStrictEqual([
        {
          checkinTime: yesterday.toISOString(),
          checkoutTime: yesterdayPlusTwoHours.toISOString(),
          forced: false,
          inProgress: false,
          name: taName,
          numHelped: 0,
        },
        {
          checkinTime: thenThreeMoreHours.toISOString(),
          checkoutTime: twelveHoursAFter.toISOString(),
          forced: true,
          inProgress: false,
          name: taName,
          numHelped: 0,
        },
        {
          checkinTime: justNow.toISOString(),
          forced: false,
          inProgress: true,
          name: taName,
          numHelped: 0,
        },
      ]);
    });
  });
});
