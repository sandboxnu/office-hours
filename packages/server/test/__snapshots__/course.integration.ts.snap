// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course Integration GET /courses/:id gets all queues that are not disabled (prof) 1`] = `
Object {
  "allowQuestions": false,
  "id": 2,
  "isDisabled": false,
  "isOpen": true,
  "isProfessorQueue": false,
  "notes": null,
  "queueSize": 0,
  "room": "room 2",
  "staffList": Array [],
}
`;

exports[`Course Integration GET /courses/:id gets all queues that are not disabled (prof) 2`] = `
Object {
  "allowQuestions": false,
  "id": 3,
  "isDisabled": false,
  "isOpen": true,
  "isProfessorQueue": false,
  "notes": null,
  "queueSize": 0,
  "room": "room 3",
  "staffList": Array [
    Object {
      "defaultMessage": null,
      "email": "user@ubc.ca",
      "firstName": "User",
      "id": 2,
      "includeDefaultMessage": true,
      "insights": Array [
        "TotalStudents",
        "TotalQuestionsAsked",
        "MedianWaitTime",
        "QuestionTypeBreakdown",
        "MostActiveStudents",
        "QuestionToStudentRatio",
        "MedianHelpingTime",
      ],
      "lastName": "Person",
      "name": "User Person",
      "password": null,
      "photoURL": null,
      "sid": null,
      "userRole": "user",
    },
  ],
}
`;

exports[`Course Integration GET /courses/:id gets all queues that are not disabled (prof) 3`] = `
Object {
  "allowQuestions": false,
  "id": 4,
  "isDisabled": false,
  "isOpen": true,
  "isProfessorQueue": true,
  "notes": null,
  "queueSize": 0,
  "room": "room 4",
  "staffList": Array [
    Object {
      "defaultMessage": null,
      "email": "user@ubc.ca",
      "firstName": "User",
      "id": 2,
      "includeDefaultMessage": true,
      "insights": Array [
        "TotalStudents",
        "TotalQuestionsAsked",
        "MedianWaitTime",
        "QuestionTypeBreakdown",
        "MostActiveStudents",
        "QuestionToStudentRatio",
        "MedianHelpingTime",
      ],
      "lastName": "Person",
      "name": "User Person",
      "password": null,
      "photoURL": null,
      "sid": null,
      "userRole": "user",
    },
  ],
}
`;

exports[`Course Integration GET /courses/:id gets office hours no queues, since no queue is happening right now 1`] = `
Object {
  "asyncQuestionDisplayTypes": Array [],
  "coordinator_email": null,
  "courseInviteCode": "invite-code",
  "crns": Array [],
  "deletedAt": null,
  "enabled": true,
  "heatmap": false,
  "icalURL": "https://calendar.google.com/calendar/ical/t6lu2pic7u9otrbpkuk26sl34g%40group.calendar.google.com/public/basic.ics",
  "id": 1,
  "name": "CS 304",
  "organizationCourse": null,
  "questionTimer": null,
  "queues": Array [],
  "sectionGroupName": "CS 304",
  "selfEnroll": false,
  "semesterId": 1,
  "timezone": "America/New_York",
  "zoomLink": null,
}
`;

exports[`Course Integration GET /courses/:id gets queues that are not disabled (TA) 1`] = `
Object {
  "allowQuestions": false,
  "id": 2,
  "isDisabled": false,
  "isOpen": true,
  "isProfessorQueue": false,
  "notes": null,
  "queueSize": 0,
  "room": "room 2",
  "staffList": Array [],
}
`;

exports[`Course Integration GET /courses/:id gets queues that are not disabled (TA) 2`] = `
Object {
  "allowQuestions": false,
  "id": 3,
  "isDisabled": false,
  "isOpen": true,
  "isProfessorQueue": false,
  "notes": null,
  "queueSize": 0,
  "room": "room 3",
  "staffList": Array [
    Object {
      "defaultMessage": null,
      "email": "user@ubc.ca",
      "firstName": "User",
      "id": 2,
      "includeDefaultMessage": true,
      "insights": Array [
        "TotalStudents",
        "TotalQuestionsAsked",
        "MedianWaitTime",
        "QuestionTypeBreakdown",
        "MostActiveStudents",
        "QuestionToStudentRatio",
        "MedianHelpingTime",
      ],
      "lastName": "Person",
      "name": "User Person",
      "password": null,
      "photoURL": null,
      "sid": null,
      "userRole": "user",
    },
  ],
}
`;

exports[`Course Integration GET /courses/:id gets queues that are not disabled (TA) 3`] = `
Object {
  "allowQuestions": false,
  "id": 4,
  "isDisabled": false,
  "isOpen": true,
  "isProfessorQueue": true,
  "notes": null,
  "queueSize": 0,
  "room": "room 4",
  "staffList": Array [
    Object {
      "defaultMessage": null,
      "email": "user@ubc.ca",
      "firstName": "User",
      "id": 2,
      "includeDefaultMessage": true,
      "insights": Array [
        "TotalStudents",
        "TotalQuestionsAsked",
        "MedianWaitTime",
        "QuestionTypeBreakdown",
        "MostActiveStudents",
        "QuestionToStudentRatio",
        "MedianHelpingTime",
      ],
      "lastName": "Person",
      "name": "User Person",
      "password": null,
      "photoURL": null,
      "sid": null,
      "userRole": "user",
    },
  ],
}
`;

exports[`Course Integration POST /courses/:id/generate_queue/:room correctly propagates notes,profq,and name 1`] = `
Object {
  "allowQuestions": true,
  "courseId": Any<Number>,
  "id": Any<Number>,
  "isDisabled": false,
  "isProfessorQueue": false,
  "notes": "example note 1",
  "room": "abcd1",
}
`;

exports[`Course Integration POST /courses/:id/generate_queue/:room correctly propagates notes,profq,and name 2`] = `
Object {
  "allowQuestions": true,
  "courseId": Any<Number>,
  "id": Any<Number>,
  "isDisabled": false,
  "isProfessorQueue": true,
  "notes": "example note 7",
  "room": "abcd2",
}
`;

exports[`Course Integration POST /courses/:id/generate_queue/:room correctly propagates notes,profq,and name 3`] = `
Object {
  "allowQuestions": true,
  "courseId": Any<Number>,
  "id": Any<Number>,
  "isDisabled": false,
  "isProfessorQueue": false,
  "notes": "ta queue",
  "room": "abcd3",
}
`;

exports[`Course Integration POST /courses/:id/ta_location/:room checks a TA into an existing queue 1`] = `
Object {
  "allowQuestions": true,
  "id": 1,
  "isDisabled": false,
  "isProfessorQueue": false,
  "notes": null,
  "room": "Online",
  "staffList": Array [
    Object {
      "defaultMessage": null,
      "email": "user@ubc.ca",
      "firstName": "User",
      "id": 1,
      "includeDefaultMessage": true,
      "insights": Array [
        "TotalStudents",
        "TotalQuestionsAsked",
        "MedianWaitTime",
        "QuestionTypeBreakdown",
        "MostActiveStudents",
        "QuestionToStudentRatio",
        "MedianHelpingTime",
      ],
      "lastName": "Person",
      "name": "User Person",
      "password": null,
      "photoURL": null,
      "sid": null,
      "userRole": "user",
    },
  ],
}
`;
